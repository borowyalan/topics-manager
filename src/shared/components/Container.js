import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Colors from '../../styles/Colors';

function commonStyle(props) {
    return `
        background-color: ${props.backgroundColor};
        color: black;
    `;
}

const DivContainer = styled.div`
    ${props => commonStyle(props)}
`;

const HeaderContainer = styled.header`
    ${props => commonStyle(props)}
`;

const SectionContainer = styled.section`
    ${props => commonStyle(props)}
`;

const FooterContainer = styled.footer`
    ${props => commonStyle(props)}
`;

const InnerContainer = styled.div`
    margin: auto;
    padding: 30px;
    max-width: 600px;
`;

const OuterContainer = props => {
    const { children, type, backgroundColor } = props;
    const allowedTags = {
        div: DivContainer,
        header: HeaderContainer,
        footer: FooterContainer,
        section: SectionContainer
    };

    if ( allowedTags[type] ) {
        const CustomContainer = allowedTags[type];
        return <CustomContainer backgroundColor={backgroundColor}>{children}</CustomContainer>;
    } else {
        return <DivContainer backgroundColor={backgroundColor}>{children}</DivContainer>;
    }
};

const Container = props => {
    const { type, backgroundColor, children, inner } = props;

    return (
        <OuterContainer type={type} backgroundColor={backgroundColor}>
            {inner === false
                ? <React.Fragment>{ children }</React.Fragment>
                : <InnerContainer>{ children }</InnerContainer>
            }
        </OuterContainer>
    );
};

Container.propTypes = {
    children: PropTypes.node.isRequired,
    backgroundColor: PropTypes.string,
    type: PropTypes.string,
    inner: PropTypes.bool
};

Container.defaultProps = {
    backgroundColor: Colors.white,
    inner: true
};


export default Container;
