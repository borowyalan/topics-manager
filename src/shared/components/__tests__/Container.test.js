import React from 'react';
import TestRenderer from 'react-test-renderer';
import { configure } from 'enzyme';
import 'jest-styled-components';
import Adapter from 'enzyme-adapter-react-16';
import Container from '../Container';

configure({ adapter: new Adapter() });

const CONTAINER_CONTENT = 'This is the container\'s content';

describe('<Container>', () => {



    it('should render as a <div /> tag when props.type is undefined', function() {
        const renderedContainer = renderContainer();
        expect(renderedContainer.type).toBe('div');
    });

    it('should render as a <div /> tag when props.type="div"', function() {
        const renderedContainer = renderContainer({ type: 'div' });
        expect(renderedContainer.type).toBe('div');
    });

    it('should render as a <header /> tag when props.type="header"', function() {
        const renderedContainer = renderContainer({ type: 'header' });
        expect(renderedContainer.type).toBe('header');
    });

    it('should render as a <footer /> tag when props.type="footer"', function() {
        const renderedContainer = renderContainer({ type: 'footer' });
        expect(renderedContainer.type).toBe('footer');
    });

    it('should render as a <section /> tag when props.type="section"', function() {
        const renderedContainer = renderContainer({ type: 'section' });
        expect(renderedContainer.type).toBe('section');
    });

    it('should have custom background-color when props.backgroundColor is defined', () => {
        const renderedContainer = renderContainer({ backgroundColor: '#990000' });
        expect(renderedContainer).toHaveStyleRule('background-color', '#990000');
    });

    it('should not have inner <div /> when props.inner={false}', () => {
        const renderedContainer = renderContainer({ inner: false, children: CONTAINER_CONTENT });
        expect(renderedContainer.children).toEqual([ CONTAINER_CONTENT ]);
    });

    it('should match snapshot', () => {
        const renderedContainer = TestRenderer.create(<Container>{CONTAINER_CONTENT}</Container>).toJSON();
        expect(renderedContainer).toMatchSnapshot();
    });

});

/**
 * Renders <Container /> with the given props
 * @param {Object} props - the props to render with
 * @return {Object} the renderedContainer component
 */
function renderContainer(props) {
    var { children = CONTAINER_CONTENT, ...rest } = props || {};
    return TestRenderer.create(<Container {...rest}>{children}</Container>).toJSON();
}
