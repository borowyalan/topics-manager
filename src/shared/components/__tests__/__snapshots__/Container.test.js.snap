// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container> should match snapshot 1`] = `
.c0 {
  background-color: #fff;
  color: black;
}

.c1 {
  margin: auto;
  padding: 30px;
  max-width: 600px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    This is the container's content
  </div>
</div>
`;
